rust/headless-client/src/lib.rs

```diff
@@ -433,49 +438,115 @@ async fn ipc_listen() -> Result<std::convert::Infallible> {
-                }
-                InternalServerMsg::OnUpdateRoutes { ipv4, ipv6 } => {
-                    tun_device.set_routes(ipv4, ipv6).await?
+impl Handler {
+    fn new(stream: IpcStream) -> Result<Self> {
+        let (rx, tx) = tokio::io::split(stream);
+        let ipc_rx = FramedRead::new(rx, LengthDelimitedCodec::new());
+        let ipc_tx = FramedWrite::new(tx, LengthDelimitedCodec::new());
+        let (cb_tx, cb_rx) = mpsc::channel(10);
+        let tun_device = tun_device_manager::TunDeviceManager::new()?;
+
+        Ok(Self {
+            callback_handler: CallbackHandler { cb_tx },
+            cb_rx,
+            connlib: None,
+            dns_controller: Default::default(),
+            ipc_rx,
+            ipc_tx,
+            last_connlib_start_instant: None,
+            tun_device,
+        })
+    }
+
+    async fn run(&mut self) -> Result<()> {
+        loop {
+            let event = {
+                // This borrows `self` so we must drop it before handling the `Event`.
+                let cb = pin!(self.cb_rx.recv());
+                match future::select(self.ipc_rx.next(), cb).await {
+                    future::Either::Left((Some(Ok(x)), _)) => Event::Ipc(
+                        serde_json::from_slice(&x)
+                            .context("Error while deserializing IPC message")?,
+                    ), // TODO: Integrate the serde_json stuff into a custom Tokio codec
+                    future::Either::Left((Some(Err(error)), _)) => Err(error)?,
+                    future::Either::Left((None, _)) => {
+                        tracing::info!("IPC client disconnected");
+                        break;
+                    }
+                    future::Either::Right((Some(x), _)) => Event::Callback(x),
+                    future::Either::Right((None, _)) => {
+                        tracing::error!("Impossible - Callback channel closed");
+                        break;
+                    }
                 }
+            };
+            match event {
+                Event::Callback(x) => self.handle_connlib_cb(x).await?,
+                Event::Ipc(msg) => self
+                    .handle_ipc_msg(msg)
+                    .context("Error while handling IPC message from client")?,
             }
         }
-        Ok::<_, anyhow::Error>(())
-    });
+        Ok(())
+    }
 
-    let mut connlib = None;
-    let callback_handler = CallbackHandler { cb_tx };
-    while let Some(msg) = rx.next().await {
-        let msg = msg?;
-        let msg: IpcClientMsg = serde_json::from_slice(&msg)?;
+    async fn handle_connlib_cb(&mut self, msg: InternalServerMsg) -> Result<()> {
+        match msg {
+            InternalServerMsg::Ipc(msg) => {
+                // The first `OnUpdateResources` marks when connlib is fully initialized
+                if let IpcServerMsg::OnUpdateResources(_) = &msg {
+                    if let Some(instant) = self.last_connlib_start_instant.take() {
```

Comment:
Touching `last_connlib_start_instant` in reaction to both incoming IPC and incoming connlib callbacks would have required a mutex or something under the previous design